contributors/design-proposals/auth/bound-service-account-tokens.md

```diff
@@ -0,0 +1,180 @@
+    account's secrets is at least as powerful as the component.
+1.  Security: JWTs are not time bound. A JWT compromised via 1 or 2, is valid
+    for as long as the service account exists. This may be mitigated with
+    service account signing key rotation but is not supported by client-go and
+    not automated by the control plane and thus is not widely deployed.
+1.  Scalability: JWTs require a Kubernetes secret per service account.
+
+# Proposal
+
+Infrastructure to support on demand token requests will be implemented in the
+core apiserver. Once this API exists, a client of the apiserver will request an
+attenuated token for it's own use. The API will enforce required attenuations,
+e.g. audience and time binding.
+
+## Token attenuations
+
+### Audience binding
+
+Tokens issued from this API will be audience bound. Audience of requested tokens
+will be bound by the `aud` claim. The `aud` claim is an array of strings
+(usually URLs) that correspond to the intended audience of the token. A
+recipient of a token is responsible for verifying that it identifies as one of
+the values in the audience claim, and should otherwise reject the token. The
+TokenReview API will support this validation.
+
+### Time binding
+
+Tokens issued from this API will be time bound. Time validity of these tokens
+will be claimed in the following fields:
+
+* `exp`: expiration time
+* `nbf`: not before
+* `iat`: issued at
+
+A recipient of a token is valid at the time that the token is presented, and
+should otherwise reject the token. The TokenReview API will support this
+validation.
+
+### Key binding
+
+Tokens issued from this API will OPTIONALLY (per the behest of the requestor)
+be key bound. The `cnf` claim will carry information that a recipient of a
+token might need to validate that the token presenter possesses a specific key.
+A recipient may require accepted tokens are key bound.
+
+## API Changes
+
+### Add `tokenrequests.authentication.k8s.io`
+
+We will add an imperative API (a la TokenReview) to the
+`authentication.k8s.io` API group:
+
+```golang
+type TokenRequest struct {
+  Spec   TokenRequestSpec
+  Status TokenRequestStatus
+}
+
+type TokenRequestSpec struct {
+  // Audience is the intendend audience of the token.
+  //
+  // see audience binding
+  Audience []string
+
+  // ValidityDuration is the intended duration of validity of the request.
+  //
+  // see time binding
+  ValidityDuration metav1.Duration
+
+  // KeyConfirmation contains information required by a recipient of a token
+  // to verify possession of a key.
+  //
+  // see key binding
+  KeyConfirmation KeyConfirmation
+}
+
+type TokenRequestStatus struct {
+  Token string
+}
+
+type KeyConfirmation struct {
+  // Structure TBD.
+  // This will not be supported in the initial implementation. This proposal
+  // will be updated later on.
+}
+
+type TokenRequestStatus struct {
+  // Token is the token data
+  Token string
```

Comment:
Can token ever be data that isn't unicode data?  I.e. do we need to deal with binary tokens?  Or will we hardcode and assume all tokens are safe for HTTP header values and for strings?

Reply:
> Can token ever be data that isn't unicode data? I.e. do we need to deal with binary tokens?

I would not expect that. Bearer tokens traditionally have a really limited character set 
```
b64token    = 1*( ALPHA / DIGIT / "-" / "." / "_" / "~" / "+" / "/" ) *"="
credentials = "Bearer" 1*SP b64token
```

TokenReviewSpec has `Token string` as well